website/docs/reference/dbt-jinja-functions/cross-database-macros.md

```diff
@@ -0,0 +1,537 @@
+---
+title: "cross-database macros"
+id: "cross-database-macros"
+---
+
+# Overview
+
+These macros make it easier for package authors (especially those writing modeling packages) to implement cross-database
+compatibility. In general, you should not need these macros in your own dbt project (unless it is a package or you otherwise desire cross-database compatibility).
+
+## All functions (alphabetical)
+- [dbt.any_value](#any_value)
+- [dbt.bool_or](#bool_or)
+- [dbt.cast_bool_to_text](#cast_bool_to_text)
+- [dbt.concat](#concat)
+- [dbt.current_timestamp](#current_timestamp)
+- [dbt.current_timestamp_in_utc](#current_timestamp_in_utc)
+- [dbt.dateadd](#dateadd)
+- [dbt.datediff](#datediff)
+- [dbt.date_trunc](#date_trunc)
+- [dbt.escape_single_quotes](#escape_single_quotes)
+- [dbt.except](#except)
+- [dbt.hash](#hash)
+- [dbt.intersect](#intersect)
+- [dbt.last_day](#last_day)
+- [dbt.length](#length)
+- [dbt.listagg](#listagg)
+- [dbt.position](#position)
+- [dbt.replace](#replace)
+- [dbt.right](#right)
+- [dbt.safe_cast](#safe_cast)
+- [dbt.split_part](#split_part)
+- [dbt.string_literal](#string_literal)
+
+## Set functions
+- [dbt.except](#except)
+- [dbt.intersect](#intersect)
+
+## String functions
+- [dbt.concat](#concat)
+- [dbt.hash](#hash)
+- [dbt.length](#length)
+- [dbt.position](#position)
+- [dbt.replace](#replace)
+- [dbt.right](#right)
+- [dbt.split_part](#split_part)
+
+## Aggregate and window functions
+- [dbt.any_value](#any_value)
+- [dbt.bool_or](#bool_or)
+- [dbt.listagg](#listagg)
+
+## Cast functions
+- [dbt.cast_bool_to_text](#cast_bool_to_text)
+- [dbt.safe_cast](#safe_cast)
+
+## Non-SQL Jinja functions
+- [dbt.escape_single_quotes](#escape_single_quotes)
+- [dbt.string_literal](#string_literal)
+
+## Timestamp functions
+- [dbt.current_timestamp](#current_timestamp)
+- [dbt.current_timestamp_in_utc](#current_timestamp_in_utc)
+
+## Date and time functions
+- [dbt.dateadd](#dateadd)
+- [dbt.datediff](#datediff)
+- [dbt.date_trunc](#date_trunc)
+- [dbt.last_day](#last_day)
+
+# Set functions
+
+## except
+__Args__:
+
+ * None
+
+`except` is one of the set operators specified ANSI SQL-92 (along with `union` and `intersect`) and is akin to [set difference](https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement).
+
+**Usage**:
+
+<File name='models/example.sql'>
+
+```sql
+{{ dbt.except() }}
```

Comment:
Maybe we should include _just_ the default "output" for each of these macros? In this case, it's hard to tell from just looking at the macro that all it does is return the string `except` (or database-specific equivalent)

Reply:
Might also help adapter authors see "at a glance" if they need to reimplement a given macro (though starting by inheriting the functional tests and seeing what fails is definitely still the Right Way To Go)

Reply:
Included the default output from the Postgres adapter for each of these macros.

Is it pretty in all cases? Nope.

We've got options for how to handle that:
- update the whitespace handling on a per-macro basis and re-generate the example output
- format the example output however we want it to look even if it is not the exact output

Do you have a preference @jtcohen6 ?

Here's the input file I used:
<details>
  <summary>/analyses/output.sql</summary>

```sql

{{ except() }}

{{ intersect() }}

{{ concat(["column_1", "column_2"]) }}
{{ concat(["year_column", "'-'" , "month_column", "'-'" , "day_column"]) }}
{{ concat(["first_part_column", "'.'" , "second_part_column"]) }}
{{ concat(["first_part_column", "','" , "second_part_column"]) }}

{{ hash("column") }}
{{ hash("'Pennsylvania'") }}

{{ length("column") }}

{{ position("substring_column", "text_column") }}
{{ position("'-'", "text_column") }}

{{ replace("string_text_column", "old_chars_column", "new_chars_column") }}
{{ replace("string_text_column", "'-'", "'_'") }}

{{ right("string_text_column", "length_column") }}
{{ right("string_text_column", "3") }}

{{ escape_single_quotes("they're") }}
{{ escape_single_quotes("ain't ain't a word") }}

select {{ string_literal("Pennsylvania") }}

{{ any_value("column_name") }}

{{ bool_or("boolean_column") }}
{{ bool_or("integer_column = 3") }}
{{ bool_or("string_column = 'Pennsylvania'") }}
{{ bool_or("column1 = column2") }}

{{ listagg(measure="column_to_agg", delimiter_text="','", order_by_clause="order by order_by_column", limit_num=10) }}

{{ cast_bool_to_text("boolean_column_name") }}
{{ cast_bool_to_text("false") }}
{{ cast_bool_to_text("true") }}
{{ cast_bool_to_text("0 = 1") }}
{{ cast_bool_to_text("1 = 1") }}
{{ cast_bool_to_text("null") }}

{{ safe_cast("column_1", api.Column.translate_type("string")) }}
{{ safe_cast("column_2", api.Column.translate_type("integer")) }}
{{ safe_cast("'2016-03-09'", api.Column.translate_type("date")) }}

{{ dateadd(datepart="day", interval=1, from_date_or_timestamp="'2016-03-09'") }}
{{ dateadd(datepart="month", interval=-2, from_date_or_timestamp="'2016-03-09'") }}

{{ datediff("column_1", "column_2", "day") }}
{{ datediff("column", "'2016-03-09'", "month") }}
{{ datediff("'2016-03-09'", "column", "year") }}

{{ date_trunc("day", "updated_at") }}
{{ date_trunc("month", "updated_at") }}
{{ date_trunc("year", "'2016-03-09'") }}

{{ last_day("created_at", "month") }}
{{ last_day("'2016-03-09'", "year") }}
```
</details>

Reply:
@dbeatty10 Genius move to use dbt to quickly & efficiently get the compiled output! Let's not talk about how I was thinking of doing this......

I think it's okay to clean up _a little_ for display in the docs. I think it's also okay if it's a bit ugly (truth in advertising). All the more reason to motivate us doing The Right Thing, which is definitely this:
> update the whitespace handling on a per-macro basis and re-generate the example output

Great opportunities for community contributors!

How will we keep this up to date going forward? Short of... the semantic layer? :)

Reply:
A handful of ideas for keeping up-to-date going forward (generally from most manual to most automated):
1. Use the same strategy as this PR:
    - `dbt init` a dbt project
    - copy-paste `/analyses/output.sql` into the project
    - execute `dbt compile --select output`
    - copy-paste from output in `target` directory into each relevant section of the docs
1. Create a git repo for the dbt project I used. Put instructions for generating the sample output in the README.
1. Build an elaborate mechanism that is mostly hands-off
1. Throw up our hands and declare the current examples in the docs are "good enough" and leave them as-is.

Option 2 (or 4!) might hit the sweet spot for us for the time being.